functions/federated_aws_rp/app.py

```diff
@@ -0,0 +1,248 @@
+    """API Gateway /redirect_uri endpoint which the user is redirected to after
+    authenticating with the identity provider
+
+    :param state: "state" query parameter returned by the identity provider
+    :param code: "code" query parameter returned by the identity provider
+    :param cookie_header: HTTP header cookies sent by the user
+    :return: An AWS API Gateway output dictionary for proxy mode
+    """
+    global discovery_document
+    if 'discovery_document' not in globals():
+        logger.debug('Fetching discovery document')
+        discovery_document = get_discovery_document(CONFIG.discovery_url)
+
+    try:
+        if 'state' is None:
+            raise AccessDenied('"state" query parameter is missing"')
+        store = get_store(cookie_header)
+        if store.get('state') != state:
+            raise AccessDenied('Invalid "state" query parameter passed')
+        if 'code' is None:
+            raise AccessDenied('"code" query parameter is missing')
+        if 'code_verifier' not in store:
+            raise AccessDenied(
+                '"code_verifier" not found in cookie')
+    except AccessDenied as e:
+        return {
+            'statusCode': 403,
+            'headers': {'Content-Type': 'text/html'},
+            'body': "Access denied : {}".format(e)}
+    token = exchange_code_for_token(
+        store['code_verifier'],
+        discovery_document['token_endpoint'],
+        code)
+    jwt.decode(
+        token=token["id_token"],
+        key=discovery_document['jwks'],
+        audience=CONFIG.client_id)
+    store['id_token'] = token['id_token']
+
+    if 'role_arn' in store:
+        return redirect_to_web_console(
+            discovery_document['jwks'],
+            store)
+    else:
+        # show role picker
+        try:
+            role_map = exchange_token_for_roles(
+                discovery_document['jwks'],
+                token["id_token"])
+        except AccessDenied as e:
+            return {
+                'statusCode': 403,
+                'headers': {'Content-Type': 'text/html'},
+                'body': "Access denied : {}".format(e)}
+        roles = {
+            "roles": []
+        }
+        for arn in role_map["roles"]:
+            account_id = arn.split(":")[4]
+            alias = role_map.get(
+                "aliases", {}).get(account_id, [account_id])[0]
+            name = arn.split(':')[5].split('/')[-1]
+
+            roles["roles"].append(
+                {
+                    "alias": alias,
+                    "arn": arn,
+                    "id": account_id,
+                    "name": name,
+                }
+            )
+
+        # Sort the list by role name
+        roles["roles"] = sorted(roles["roles"], key=lambda x: x['name'])
+        response = {
+            'statusCode': 200,
+            'headers': {
+                'Content-Type': 'text/html',
+                'Set-Cookie': '{}={}; Secure; HttpOnly; Max-Age=3600'.format(
```

Comment:
I think you can probably use `SameSite` here, to protect you against CSRF attacks.

Reply:
Sure, I think `SameSite=strict` should work as the one case where a user comes to the RP from another site, when they're session needs to be refreshed, the only two pieces of data that the RP uses are `referer` and query parameters

Reply:
Fixed in 58eead07041781871f85186762fc4050c9248dd3